### PR TITLE
Refs [TASK][34938] Handle event Notification

### PR DESCRIPTION
**Pull request by manhhung23399**
## Related Tickets
- [#Task 34938: Handle event Notification](https://edu-redmine.sun-asterisk.vn/issues/34938)
## WHAT
- Create notification
- Handle Click notification
## Evidence (Screenshot or Video)
image

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/example/moviews/screen/home/adapter/ViewPagerAdapter.kt [hunk 1]
```diff
@@ -14,10 +14,14 @@ import kotlinx.android.synthetic.main.item_slide.view.*
...
-
+    private val limit = 5
     private val movies = mutableListOf<Movie>()
 
-    override fun getCount(): Int = movies.size
+    override fun getCount(): Int {
+        return if (movies.size > limit) {
+            limit
+        } else movies.size
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/screen/home/adapter/ViewPagerAdapter.kt]: dùng min(size, 5)

File: app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt [hunk 2]
```diff
@@ -1,29 +1,48 @@
...
+import android.content.SharedPreferences
 import com.example.moviews.R
 import com.example.moviews.base.BaseFragment
+import com.example.moviews.utils.Notification
 import kotlinx.android.synthetic.main.fragment_information.*
 
-class InformationFragment : BaseFragment() {
 
+class InformationFragment : BaseFragment() {
+    private var check: Boolean? = null
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt]: check gì ? đặt tên rõ ràng

File: app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt [hunk 3]
```diff
@@ -1,29 +1,48 @@
...
+                if (isChecked) {
+                    Notification.enable(it)
+                } else {
+                    Notification.disable(it)
+                }
+                val preferences: SharedPreferences? =
+                    it.getSharedPreferences(PREF_SETTING, MODE_PRIVATE)
+                val editor = preferences?.edit()
+                editor?.putBoolean(PREF_SWITCH, isChecked)
+                editor?.apply()
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt:38]: Tại sao lại xử lý data trên View ?

File: app/src/main/java/com/example/moviews/service/ServiceNotify.kt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+class ServiceNotify : JobIntentService() {
+
+    override fun onHandleWork(intent: Intent) {
+        val movieRepository = RepositoryUtils.getMovieRepository(applicationContext)
+        movieRepository.getMovies(
+            Constant.BASE_TRENDING,
+            object : OnLoadDataCallback<MutableList<Movie>> {
+                override fun onSuccess(data: MutableList<Movie>) {
+                    val random = Random().nextInt(data.size)
+                    showNotification(data[random])
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/service/ServiceNotify.kt]: Collection hỗ trợ sẵn random 
```suggestion
                    showNotification(data.random())
```

File: app/src/main/java/com/example/moviews/service/ServiceNotify.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        val collapsedView = RemoteViews(
+            packageName,
+            R.layout.layout_notification_collapsed
+        )
+        val expandedView = RemoteViews(
+            packageName,
+            R.layout.layout_notification_expanded
+        )
+        expandedView.setTextViewText(R.id.textTitleMovieNotify, movie.title)
+        val intent = Intent(this, MainActivity::class.java)
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/service/ServiceNotify.kt]: conventions 2.11

File: app/src/main/java/com/example/moviews/service/ServiceNotify.kt [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            R.id.imageNotify,
+            expandedView,
+            builder,
+            NOTIFICATION_ID
+        )
+        Glide
+            .with(this)
+            .asBitmap()
+            .load(BaseUrl.baseUrlImage(movie.poster))
+            .into(target)
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/service/ServiceNotify.kt]: hàm này quá dài, tách hàm ra cho clear

File: app/src/main/java/com/example/moviews/service/ServiceNotify.kt [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import com.example.moviews.MainActivity
+import com.example.moviews.R
+import com.example.moviews.data.OnLoadDataCallback
+import com.example.moviews.data.model.Movie
+import com.example.moviews.repository.RepositoryUtils
+import com.example.moviews.utils.BaseUrl
+import com.example.moviews.utils.Constant
+import java.util.*
+
+class ServiceNotify : JobIntentService() {
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/service/ServiceNotify.kt]:
```suggestion
class NotificationService : JobIntentService() {
```

File: app/src/main/java/com/example/moviews/utils/Notification.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+import android.app.AlarmManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import com.example.moviews.receiver.NotifyReceiver
+import java.util.*
+
+object Notification {
+    fun enable(context: Context) {
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/utils/Notification.kt:11]: sao không truyền vào constructor

File: app/src/main/java/com/example/moviews/utils/Notification.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                intendedTime,
+                AlarmManager.INTERVAL_DAY,
+                pendingIntent
+            )
+        }
+    }
+
+    fun disable(context: Context) {
+        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
+        val intent = Intent(context, NotifyReceiver::class.java)
```
Comment (huytq-2422) [inline app/src/main/java/com/example/moviews/utils/Notification.kt]: conventions 2.11

File: app/src/main/res/layout/layout_notification_expanded.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (huytq-2422) [inline app/src/main/res/layout/layout_notification_expanded.xml:2]: ConstraintLayout

File: app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt [hunk 11]
```diff
@@ -1,29 +1,48 @@
...
     }
 
     override fun initData() {
+        val preferences: SharedPreferences? =
+            context?.getSharedPreferences(PREF_SETTING, MODE_PRIVATE)
+        check = preferences?.getBoolean(PREF_SWITCH, false)
     }
 
     override fun initEvents() {
+        switchNotification.isChecked = check == true
```
Comment (datdt-1935) [inline app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt]:
```suggestion
        switchNotification.isChecked = check
```

File: app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt (hunk 3 above)
Comment (manhhung23399) [reply app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt:38]: chỗ này e bắt sự kiện onCheckedchange của switch để put vào sharedpreferences như này ko đc hả a?

File: app/src/main/java/com/example/moviews/utils/Notification.kt (hunk 8 above)
Comment (manhhung23399) [reply app/src/main/java/com/example/moviews/utils/Notification.kt:11]: lớp object thì k đc truyền ạ

File: app/src/main/res/layout/layout_notification_expanded.xml (hunk 10 above)
Comment (manhhung23399) [reply app/src/main/res/layout/layout_notification_expanded.xml:2]: remoteViews ko cho dùng constraint a

Comment (manhhung23399) [reply app/src/main/res/layout/layout_notification_expanded.xml:2]: image

File: app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt (hunk 11 above)
Comment (manhhung23399) [reply app/src/main/java/com/example/moviews/screen/information/InformationFragment.kt]: chỗ này bỏ đi k đc a ạ, tại bên trên e khai báo biến check = null rồi

**Review by datdt-1935 — APPROVED** [review]

**Pull request merged**